### PR TITLE
TransactionBuilder Cleanup

### PR DESCRIPTION
**Pull request by Maithem**
## Overview
Proper implementation of a TransactionBuilder.

Why should this be merged: cleanup required to add transaction validation. 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java [hunk 1]
```diff
@@ -55,7 +55,7 @@
             new HashSet<>();
 
 
-    OptimisticTransactionalContext(TransactionBuilder builder) {
+    OptimisticTransactionalContext(Transaction builder) {
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: please rename builder variable

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java [hunk 2]
```diff
@@ -271,7 +271,7 @@ public long getConflictSetAndCommit(ConflictSetInfo conflictSet) {
         // Write to the transaction stream if transaction logging is enabled
         Set<UUID> affectedStreamsIds = new HashSet<>(getWriteSetInfo().getWriteSet().getEntryMap().keySet());
 
-        if (this.builder.runtime.getObjectsView().isTransactionLogging()) {
+        if (this.transaction.runtime.getObjectsView().isTransactionLogging()) {
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: would be nice to provide `this.transaction.isTransactionLogging()`

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java [hunk 3]
```diff
@@ -323,7 +323,7 @@ protected void tryCommitAllProxies() {
         // First, get the committed entry
         // in order to get the backpointers
         // and the underlying SMREntries.
-        ILogData committedEntry = this.builder.getRuntime()
+        ILogData committedEntry = this.transaction.getRuntime()
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: Why not to pass `CorfuRuntime` in a constructor as a parameter?

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/SnapshotTransactionalContext.java [hunk 4]
```diff
@@ -32,8 +32,8 @@
     @Getter
     private Set<ICorfuSMRProxyInternal> modifiedProxies = ImmutableSet.of();
 
-    public SnapshotTransactionalContext(TransactionBuilder builder) {
-        super(builder);
+    public SnapshotTransactionalContext(Transaction transaction) {
+        super(transaction);
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/SnapshotTransactionalContext.java]: Please pass `CorfuRuntime` as a parameter, don't use transaction.getRuntime

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/Transaction.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    public void begin() {
+        //TODO(Maithem) move timestamp validation here
+        TransactionalContext.newContext(type.get.apply(this));
+    }
+
+    public static TransactionBuilder builder() {
+        return new TransactionBuilder();
+    }
+
+    public static class TransactionBuilder {
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/Transaction.java]: Please use lombok `@Builder`, this code is boilerplate

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java (hunk 2 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: transaction logging is not per transaction, its runtime wide.

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/SnapshotTransactionalContext.java (hunk 4 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/object/transactions/SnapshotTransactionalContext.java]: Transactions are tied to a particular runtime.

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java (hunk 1 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java (hunk 2 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: Its much clean now. Thanks!

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java (hunk 3 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: Just renaming, not changing the logic of tryCommitAllProxies

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/SnapshotTransactionalContext.java (hunk 4 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/object/transactions/SnapshotTransactionalContext.java]: Just renaming, not changing the logic of tryCommitAllProxies

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/Transaction.java (hunk 5 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/object/transactions/Transaction.java]: Done. Thanks for the advice.

**Comment by Maithem** [comment]
@annym I'm trying to push this pr and #1567 should I wait for the cron job to complete the injection-framework or should i invoke it manually

**Comment by Maithem** [comment]
@zalokhan yes.

**Pull request merged**